### ISSUE TITLE
app-id: Revoke a user-id mapping

### ISSUE THREAD

**Issue by jsok** [hashicorp/vault#191]
So following the example in the docs:

```
vault write auth/app-id/map/app-id/foo value=root display_name=foo
vault write auth/app-id/map/user-id/bar value=foo
```

Is it possible to revoke the user-id `bar` so I can no longer login with `{"app_id": "foo", "user_id": "bar"}`?

**Comment by sethvargo** [comment]
Hi @jsok,

I believe you can use the `revoke` command here:

```
vault revoke auth/app-id/map/user-id/bar
```

Or, if you wanted to disable auth for that app entirely:

```
vault revoke -prefix auth/app-id
```

Does that answer your question?

**Comment by zerosignal0** [comment]
If I revoke an app-id / user-id it doesn't stop them from logging back in.  Is there a way to delete an app-id / user-id?  

I have tried to find a way but it doesn't seem to work currently, as I continue to face 500 errors.

**Comment by jsok** [comment]
+1 @zerosignal0, I'm seeing the same behaviour.

Below is my session output:
```
> vault auth-enable app-id
Successfully enabled 'app-id' at 'app-id'!
> vault write auth/app-id/map/app-id/foo value=root display_name=foo
Success! Data written to: auth/app-id/map/app-id/foo
> vault write auth/app-id/map/user-id/bar value=foo
Success! Data written to: auth/app-id/map/user-id/bar
> curl --silent --data '{"app_id": "foo", "user_id": "bar"}' "$VAULT_ADDR/v1/auth/app-id/login" | python -m json.tool
{
    "auth": {
        "client_token": "63b1354b-cee0-826d-d4af-0a139a2b0c9c",
        "lease_duration": 0,
        "metadata": {
            "app-id": "sha1:0beec7b5ea3f0fdbc95d0dd47f3c5bc275da8a33",
            "user-id": "sha1:62cdb7020ff920e5aa642c3d4066950dd1f01f4d"
        },
        "policies": [
            "root"
        ],
        "renewable": false
    },
    "data": null,
    "lease_duration": 0,
    "lease_id": "",
    "renewable": false
}
> vault revoke auth/app-id/map/user-id/bar
Key revoked with ID 'auth/app-id/map/user-id/bar'.
> curl --silent --data '{"app_id": "foo", "user_id": "bar"}' "$VAULT_ADDR/v1/auth/app-id/login" | python -m json.tool
{
    "auth": {
        "client_token": "b78ff7e5-e0d2-d3bc-0512-e80b8f0dc373",
        "lease_duration": 0,
        "metadata": {
            "app-id": "sha1:0beec7b5ea3f0fdbc95d0dd47f3c5bc275da8a33",
            "user-id": "sha1:62cdb7020ff920e5aa642c3d4066950dd1f01f4d"
        },
        "policies": [
            "root"
        ],
        "renewable": false
    },
    "data": null,
    "lease_duration": 0,
    "lease_id": "",
    "renewable": false
}
```

**Comment by armon** [comment]
@jsok @zerosignal0 I can confirm this is a bug. Those endpoints should support delete, but do not currently. I've labeled it as such, should be an easy fix.

**Comment by jsok** [comment]
Thanks @armon. BTW is there a way to map app and user IDs via the HTTP API? It doesn't seems to be exposed in the code (only `/login` so far).

**Comment by armon** [comment]
@jsok It looks like you figured out the mapping via the HTTP API! It is indeed exposed

**Comment by ZeeShen** [comment]
@armon v0.5.3 cannot revoke user-id, too.
My operation:

```
root@vault-stg1:~# vault -v
Vault v0.5.3
root@vault-stg1:~# vault write secret/test value=hello
Success! Data written to: secret/test

root@vault-stg1:~# vault write auth/app-id/map/app-id/foo value=root display_name=foo
Success! Data written to: auth/app-id/map/app-id/foo
root@vault-stg1:~# vault write auth/app-id/map/user-id/bar value=foo
Success! Data written to: auth/app-id/map/user-id/bar
```

After mapping, login & read secret is ok
```
root@vault-stg1:~# curl -d '{"app_id": "foo", "user_id":"bar"}' "http://localhost:8087/v1/auth/app-id/login"
{"lease_id":"","renewable":false,"lease_duration":0,"data":null,"warnings":null,"auth":{"client_token":"203edcfe-3a5a-17e8-a56f-ea42d6bd5e4b","accessor":"c95aec5d-4a48-97de-8c94-0231dfb46fef","policies":["root"],"metadata":{"app-id":"sha1:0beec7b5ea3f0fdbc95d0dd47f3c5bc275da8a33","user-id":"sha1:62cdb7020ff920e5aa642c3d4066950dd1f01f4d"},"lease_duration":0,"renewable":true}}

root@vault-stg1:~# curl -XGET -H "X-Vault-Token: 203edcfe-3a5a-17e8-a56f-ea42d6bd5e4b" "http://localhost:8087/v1/secret/test"
{"lease_id":"","renewable":false,"lease_duration":2592000,"data":{"value":"hello"},"warnings":null,"auth":null}
```

After revoke, still can login & read secret with either new or old token
```
root@vault-stg1:~# vault revoke auth/app-id/map/user-id/bar
Key revoked with ID 'auth/app-id/map/user-id/bar'.
root@vault-stg1:~# curl -XGET -H "X-Vault-Token: 203edcfe-3a5a-17e8-a56f-ea42d6bd5e4b" "http://localhost:8087/v1/secret/test"
{"lease_id":"","renewable":false,"lease_duration":2592000,"data":{"value":"hello"},"warnings":null,"auth":null}

root@vault-stg1:~# curl -d '{"app_id": "foo", "user_id":"bar"}' "http://localhost:8087/v1/auth/app-id/login"
{"lease_id":"","renewable":false,"lease_duration":0,"data":null,"warnings":null,"auth":{"client_token":"bdccad6a-0a69-5b73-360d-39f85a13f724","accessor":"49fcde9b-1bf7-a2b7-b8bf-7d28567082c3","policies":["root"],"metadata":{"app-id":"sha1:0beec7b5ea3f0fdbc95d0dd47f3c5bc275da8a33","user-id":"sha1:62cdb7020ff920e5aa642c3d4066950dd1f01f4d"},"lease_duration":0,"renewable":true}}

root@vault-stg1:~# curl -XGET -H "X-Vault-Token: bdccad6a-0a69-5b73-360d-39f85a13f724" "http://localhost:8087/v1/secret/test"
{"lease_id":"","renewable":false,"lease_duration":2592000,"data":{"value":"hello"},"warnings":null,"auth":null}
```

And I tried to delete mapping, cannot login  but still can read secret. 
```
root@vault-stg1:~# vault delete auth/app-id/map/user-id/bar
Success! Deleted 'auth/app-id/map/user-id/bar'

root@vault-stg1:~# curl -XGET -H "X-Vault-Token: 203edcfe-3a5a-17e8-a56f-ea42d6bd5e4b" "http://localhost:8087/v1/secret/test"
{"lease_id":"","renewable":false,"lease_duration":2592000,"data":{"value":"hello"},"warnings":null,"auth":null}
root@vault-stg1:~# curl -XGET -H "X-Vault-Token: bdccad6a-0a69-5b73-360d-39f85a13f724" "http://localhost:8087/v1/secret/test"
{"lease_id":"","renewable":false,"lease_duration":2592000,"data":{"value":"hello"},"warnings":null,"auth":null}
```

I saw you closed this issue at 2015 and `v0.5.3` released this year. So could you please help to check that this revoke mapping is working or not? And I think that after revoke, neither login nor read secret can be done, is that right?
